crates/p2p/examples/peer.rs

```diff
@@ -0,0 +1,180 @@
+#![deny(rust_2018_idioms)]
+
+use std::sync::Arc;
+use std::{path::Path, time::Duration};
+
+use clap::Parser;
+use libp2p::Multiaddr;
+use libp2p::{identity::Keypair, PeerId};
+use p2p::Peers;
+use p2p_proto as proto;
+use proto::sync::{BlockBodies, StateDiffs};
+use serde_derive::Deserialize;
+use stark_hash::Felt;
+use tokio::sync::RwLock;
+use zeroize::Zeroizing;
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[clap(long, value_parser, env = "IDENTITY_CONFIG_FILE")]
+    identity_config_file: Option<std::path::PathBuf>,
+    #[clap(long, value_parser, env = "LISTEN_ON")]
+    listen_on: Multiaddr,
+    #[clap(long, value_parser, env = "BOOTSTRAP_ADDRESSES")]
+    bootstrap_addresses: Vec<Multiaddr>,
+    #[clap(long)]
+    emit_events: bool,
+}
+
+#[derive(Clone, Deserialize)]
+struct IdentityConfig {
+    pub private_key: String,
+}
+
+impl IdentityConfig {
+    pub fn from_file(path: &Path) -> anyhow::Result<Self> {
+        Ok(serde_json::from_str(&std::fs::read_to_string(path)?)?)
+    }
+}
+
+impl zeroize::Zeroize for IdentityConfig {
+    fn zeroize(&mut self) {
+        self.private_key.zeroize()
+    }
+}
+
+const PERIODIC_STATUS_INTERVAL: Duration = Duration::from_secs(30);
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    if std::env::var_os("RUST_LOG").is_none() {
+        std::env::set_var("RUST_LOG", "info");
+    }
+
+    setup_tracing();
+
+    let args = Args::parse();
+    let keypair = match &args.identity_config_file {
+        Some(path) => {
+            let config = Zeroizing::new(IdentityConfig::from_file(path.as_path())?);
+            let private_key = Zeroizing::new(base64::decode(config.private_key.as_bytes())?);
+            Keypair::from_protobuf_encoding(&private_key)?
+        }
+        None => {
+            tracing::info!("No private key configured, generating a new one");
+            Keypair::Ed25519(libp2p::identity::ed25519::Keypair::generate())
+        }
+    };
+
+    let peer_id = keypair.public().to_peer_id();
+    tracing::info!(%peer_id, "Starting up");
+
+    let peers: Arc<RwLock<Peers>> = Arc::new(RwLock::new(Default::default()));
+    let (mut p2p_client, mut p2p_events, p2p_main_loop) =
+        p2p::new(keypair, peers.clone(), PERIODIC_STATUS_INTERVAL);
+
+    let _p2p_task = tokio::task::spawn(p2p_main_loop.run());
+
+    p2p_client.start_listening(args.listen_on).await?;
+
+    for bootstrap_address in args.bootstrap_addresses {
+        let peer_id = bootstrap_address
+            .iter()
+            .find_map(|p| match p {
+                libp2p::multiaddr::Protocol::P2p(h) => PeerId::from_multihash(h).ok(),
+                _ => None,
+            })
+            .ok_or_else(|| anyhow::anyhow!("Boostrap addresses must inlcude peer ID"))?;
```

Comment:
Is this outlined somewhere in the docs? Afaik using the rust implementation the peer id can be omitted (just checked in the bootstrap test). Are the other implementations are reason https://github.com/libp2p/go-libp2p/issues/1040?

Reply:
I don't remember clearly. However, independently of whether or not it's technically required I think requiring the peer id is a good idea. Without that you can't be sure you're connecting to the node you're expecting?

Reply:
Yeah it makes sense.